File: Chapter 2/Section 4/FenwickTree.java
import java.util.*;

class FenwickTree {
     private ArrayList<Integer> ft;

     private int LSOne(int s) {
          return (s & (-s));
     }

     public FenwickTree(int n) {
          ft = new ArrayList<>();
          for (int i = 0; i <= n; i++)
               ft.add(0);
     }

     public int rsq(int j) {
          int sum = 0;
          for (; j > 0; j -= LSOne(j))
               sum += ft.get(j);
          return sum;
     }

     public int rsq(int i, int j) {
          return rsq(j) - rsq(i - 1);
     }

     public void update(int i, int v) {
          for (; i < (int) ft.size(); i += LSOne(i))
               ft.set(i, ft.get(i) + v);
     }
}

File: Chapter 2/Section 4/FenwickTree.cpp
#include <bits/stdc++.h>

#define LSOne(S) ((S) & -(S))

using namespace std;

typedef long long ll;
typedef vector<ll> vll;

class FenwickTree {
private: vll ft;
public:
     FenwickTree(int n) {
          ft.assign(n + 1, 0);
     }
     void build(const vll &f) {
          int n = (int) f.size() - 1;
          ft.assign(n + 1, 0);
          for (int i = 1; i <= n; i++) {
               ft[i] += f[i];
               if (i + LSOne(i) <= n) {
                    ft[i + LSOne(i)] += ft[i];
               }
          }
     }
     FenwickTree(const vll &f) {
          build(f);
     }
     ll rsq(int a) {
          int sum = 0;
          for (; a; a -= LSOne(a)) sum += ft[a];
          return sum;
     }
     ll rsq(int a, int b) {
          return rsq(b) - ((a == 1) ? 0 : rsq(a));
     }
     void update(int a, int val) {
          for (; a < (int) ft.size(); a += LSOne(a)) ft[a] += val;
     }
};
